system/lib/compiler-rt/emscripten_setjmp.c

```diff
@@ -73,33 +74,54 @@ void emscripten_longjmp(uintptr_t env, int val) {
 #endif
 
 #ifdef __USING_WASM_SJLJ__
-
 struct __WasmLongjmpArgs {
   void *env;
   int val;
 };
+#endif
 
-thread_local struct __WasmLongjmpArgs __wasm_longjmp_args;
+// jmp_buf should have large enough size and alignment to contain
+// this structure.
+struct jmp_buf_impl {
+  void* func_invocation_id;
+  uint32_t label;
+#ifdef __USING_WASM_SJLJ__
+  struct __WasmLongjmpArgs arg;
+#endif
+};
 
-// llvm uses `1` for the __c_longjmp tag.
-// See https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/CodeGen/WasmEHFuncInfo.h
-#define C_LONGJMP 1
+void __wasm_setjmp(void* env, uint32_t label, void* func_invocation_id) {
+  struct jmp_buf_impl* jb = env;
+  assert(label != 0);                 // ABI contract
+  assert(func_invocation_id != NULL); // sanity check
+  jb->func_invocation_id = func_invocation_id;
+  jb->label = label;
+}
 
-// Wasm EH allows us to throw and catch multiple values, but that requires
-// multivalue support in the toolchain, whch is not reliable at the time.
-// TODO Consider switching to throwing two values at the same time later.
-void __wasm_longjmp(void *env, int val) {
-  __wasm_longjmp_args.env = env;
+uint32_t __wasm_setjmp_test(void* env, void* func_invocation_id) {
+  struct jmp_buf_impl* jb = env;
+  assert(jb->label != 0);             // ABI contract
+  assert(func_invocation_id != NULL); // sanity check
+  if (jb->func_invocation_id == func_invocation_id) {
+    return jb->label;
+  }
+  return 0;
+}
+
+#ifdef __USING_WASM_SJLJ__
+void __wasm_longjmp(void* env, int val) {
+  struct jmp_buf_impl* jb = env;
+  struct __WasmLongjmpArgs* arg = &jb->arg;
   /*
-￼  * C standard:
-￼  *   The longjmp function cannot cause the setjmp macro to return
-￼  *   the value 0; if val is 0, the setjmp macro returns the value 1.
-￼  */
+   * C standard says:
+   * The longjmp function cannot cause the setjmp macro to return
+   * the value 0; if val is 0, the setjmp macro returns the value 1.
+   */
   if (val == 0) {
     val = 1;
   }
-  __wasm_longjmp_args.val = val;
-  __builtin_wasm_throw(C_LONGJMP, &__wasm_longjmp_args);
+  arg->env = env;
+  arg->val = val;
+  __builtin_wasm_throw(1, arg); /* 1 == C_LONGJMP */
```

Comment:
Is this code going to work both before and after the llvm change?

Reply:
> Is this code going to work both before and after the llvm change?

i expect so.
at least it passed ./test/runner.py 'other.test_setjmp*' and ./test/runner.py 'other.test_setjmp*'.
do you have specific concerns?

Reply:
I guess I don't see how this same function can work both before and after the llvm change.   Won't llvm generate different `env` arguments before and after your change?   Is the layout of `env` not changing with your llvm change?

Can you also run `./test/runner core0.test_setjmp*`?

Reply:
it passed `./test/runner core0.test_setjmp*` as well. (it was what i meant. repeating two `other.test_setjmp*` was a copy-and-paste mistake.)

the original jmp_buf was large enough to store &jb->arg. it was just unused before this change.